### PR TITLE
[`pipeline`] Fix str device issue

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Addresses: https://github.com/huggingface/transformers/pull/24140#issuecomment-1584617146 

Currently passing `device="cuda"` is not supported when creating a pipeline. 
This is because `torch.cuda.set_device(self.device)` expects the device to have an explicit index. The fix is to create an indexed device when initializing a pipeline with a str device

Handy reproducible snippet:

```python
from transformers import pipeline

# this works
pipe = pipeline("text-generation", device=0)
pipe("Hello")

# this works
pipe = pipeline("text-generation", device="cuda:0")
pipe("Hello")

# this fails
pipe = pipeline("text-generation", device="cuda")
pipe("Hello")
```

cc @amyeroberts @Narsil

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

Just some more comments on making it a bit more robust. They're suggestions, so up to you if you want to add.

File: src/transformers/pipelines/base.py [hunk 1]
```diff
@@ -793,6 +793,8 @@ def __init__(
             if isinstance(device, torch.device):
                 self.device = device
             elif isinstance(device, str):
+                if device == "cuda":
+                    device = f"cuda:{torch.cuda.current_device()}"
                 self.device = torch.device(device)
             elif device < 0:
                 self.device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/base.py:800]: https://pytorch.org/docs/stable/generated/torch.cuda.set_device.html

`set_device` seems unsupported, so I'm unsure about `current_device()` usage.

```
torch.device("cuda")
```
Works though, what's the issue ?

**Comment by Narsil** [comment]
Also 

```
python -c 'from transformers import pipeline; pipe = pipeline(model="gpt2", device="cuda")'
```

Works on `main`.. So I'm not sure what's the issue

**Comment by Narsil** [comment]
Can we remove the `set_device` instead then ? Seems better:

```patch
diff --git a/src/transformers/pipelines/base.py b/src/transformers/pipelines/base.py
index 510c07cf5..b5975d081 100644
--- a/src/transformers/pipelines/base.py
+++ b/src/transformers/pipelines/base.py
@@ -901,10 +901,8 @@ class Pipeline(_ScikitCompat):
             with tf.device("/CPU:0" if self.device == -1 else f"/device:GPU:{self.device}"):
                 yield
         else:
-            if self.device.type == "cuda":
-                torch.cuda.set_device(self.device)
-
-            yield
+            with torch.cuda.device(self.device):
+                yield
```

**Comment by younesbelkada** [comment]
I am happy to revert some of the changes I proposed and add yours, it looks much better. However I have few questions 
1- is it ok to call that context manager if `self.device` is CPU? I think we need a check on top of that to make sure we're not on CPU (similarly as what we had before)

```python
import torch
device = torch.device("cpu")

with torch.cuda.device(device):
    print(torch.randn(1))
```
Throws:
```bash
    raise ValueError('Expected a cuda device, but got: {}'.format(device))
ValueError: Expected a cuda device, but got: cpu
```

2- I am not sure but I think the `with device` context manager is only available since PT2.0 no?

**Comment by Narsil** [comment]
> 2- I am not sure but I think the with device context manager is only available since PT2.0 no?

I don't know, all those are very good questions for which I don't have the answer to. I just know that now `set_device` is strongly discouraged so it's probably the source of our issues.

**Comment by younesbelkada** [comment]
Thanks ! 
I can confirm the context manager doesn't work for PT==1.9 which is [should be supported by us](https://github.com/huggingface/transformers/blob/4c6e42958951ca66a6b498b1afce8d8ad4ac2274/setup.py#L178): 

```python
Traceback (most recent call last):
  File "scratch.py", line 203, in <module>
    with torch.device(device):
AttributeError: __enter__
```

Therefore I just added some changes to ensure backward compatibility with older PT versions. WDYT?

File: src/transformers/pipelines/base.py [hunk 2]
```diff
@@ -793,11 +793,16 @@ def __init__(
             if isinstance(device, torch.device):
                 self.device = device
             elif isinstance(device, str):
+                if device == "cuda":
```
Comment (younesbelkada) [inline src/transformers/pipelines/base.py:796]: Maybe?
```suggestion
                if device == "cuda" and not hasattr(torch.device, "__enter__"):
```

File: src/transformers/pipelines/base.py [hunk 3]
```diff
@@ -902,10 +907,14 @@ def device_placement(self):
             with tf.device("/CPU:0" if self.device == -1 else f"/device:GPU:{self.device}"):
                 yield
         else:
-            if self.device.type == "cuda":
-                torch.cuda.set_device(self.device)
-
-            yield
+            if hasattr(torch.device, "__enter__"):
```
Comment (amyeroberts) [inline src/transformers/pipelines/base.py:910]: nit: we typically check for compatibility with flags like `is_torch_greater_or_equal_than_2_0` in [pytorch_utils.](https://github.com/huggingface/transformers/blob/1815d1865e61df3be1cbe37ee8b78f6017b0b1e2/src/transformers/pytorch_utils.py#L31). It's a bit cleaner than checking for a private attribute and is clearer for the reader what's being checked

File: src/transformers/pipelines/base.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/pipelines/base.py:796]: Just double checking, if this condition is true, does the line below run OK?

`self.device = torch.device(device)`

**Review by amyeroberts — APPROVED** [review]
I think the updates look good 👍

File: src/transformers/pipelines/base.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/pipelines/base.py:910]: Ok I can confirm it has been introduced on PT>=2.0.0

**Comment by Narsil** [comment]
May I attempt a different thing ?

I think the fix is correct, but I'm wondering if simply relying on `torch.cuda.device` context manager couldn't help remove the need for the compat layer.

**Comment by younesbelkada** [comment]
Sure yes!

**Comment by Narsil** [comment]
Cannot push

```patch
diff --git a/src/transformers/pipelines/base.py b/src/transformers/pipelines/base.py
index 626d33a3d..ee117e62a 100644
--- a/src/transformers/pipelines/base.py
+++ b/src/transformers/pipelines/base.py
@@ -50,7 +50,6 @@ if is_torch_available():
     from torch.utils.data import DataLoader, Dataset

     from ..models.auto.modeling_auto import AutoModel
-    from ..pytorch_utils import is_torch_greater_or_equal_than_2_0

     # Re-export for backward compatibility
     from .pt_utils import KeyDataset
@@ -794,16 +793,11 @@ class Pipeline(_ScikitCompat):
             if isinstance(device, torch.device):
                 self.device = device
             elif isinstance(device, str):
-                if device == "cuda" and not is_torch_greater_or_equal_than_2_0:
-                    # for backward compatiblity if using `set_device` and `cuda`
-                    device = f"cuda:{torch.cuda.current_device()}"
                 self.device = torch.device(device)
             elif device < 0:
                 self.device = torch.device("cpu")
-            elif isinstance(device, int):
-                self.device = torch.device(f"cuda:{device}")
             else:
-                raise ValueError(f"Device type not supported. Got {device}")
+                self.device = torch.device(f"cuda:{device}")
         else:
             self.device = device if device is not None else -1
         self.torch_dtype = torch_dtype
@@ -908,13 +902,10 @@ class Pipeline(_ScikitCompat):
             with tf.device("/CPU:0" if self.device == -1 else f"/device:GPU:{self.device}"):
                 yield
         else:
-            if is_torch_greater_or_equal_than_2_0:
-                with torch.device(self.device):
+            if self.device.type == "cuda":
+                with torch.cuda.device(self.device):
                     yield
-            # for backward compatibility
             else:
-                if self.device.type == "cuda":
-                    torch.cuda.set_device(self.device)
                 yield
```

**Comment by younesbelkada** [comment]
Hi @Narsil 
I am not sure if `with torch.cuda.device(self.device):` is supported for torch<2.0 

https://pytorch.org/tutorials/recipes/recipes/changing_default_device.html

Maybe we should merge this PR for now to unlock also @thomasw21 . what do you think?

**Comment by thomasw21** [comment]
I don't think we're blocked by this. 

> And torch.device("cpu") ... well it's the default there's no need to context manage it.

Not sure of the context of this sentence, but we're overriding the default to `cuda`, so having a context manager to switch back to `cpu` makes sense to me.

**Comment by Narsil** [comment]
https://pytorch.org/docs/1.9.0/generated/torch.cuda.device.html?highlight=torch%20cuda%20device#torch.cuda.device

It is supported from 1.9.0+, at least in the docs.

**Comment by younesbelkada** [comment]
Great ! agreed with those changes

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

**Pull request merged**